### PR TITLE
fix: prevent silent failure when DOM storage quota exceeded

### PR DESCRIPTION
**Pull request by jacobq**
#### Description of Change
This PR modifies [`dom_storage_limits.patch`](https://github.com/electron/electron/blob/ba85d4c3bbbaddb53795340053c273dfdf51aca1/patches/chromium/dom_storage_limits.patch) so that instead of (incompletely / incorrectly) circumventing the enforcement of the DOM storage quota, it is simply increased (from 10MiB to 100MiB). The primary motivation for this change is that the previous patch (see PR #15596) does not actually work properly ([example](https://github.com/jacobq/electron-quick-start/tree/storageLimit)). 

As described in the [updated patch notes](https://github.com/electron/electron/pull/20899/files#diff-557eb2fa2d8baedff18ed8d626af55c4), bypassing the [maximum size enforcement in
`StorageAreaImpl::Put`](https://github.com/chromium/chromium/blob/62521aaefdef5b345dd470f14eef6e44732ccf23/content/browser/dom_storage/storage_area_impl.cc#L306-L315) partially resolves the problem but can lead to undesirable side-effects (e.g. instability / poor performance). 

**Cc:**
@deepak1556 
@alexeykuzmin 
@nornagon 

#### Checklist
- [X] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [X] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [X] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [X] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: DOM storage quota enforcement re-enabled, but limit changed from 10MiB to 100MiB

### CODE REVIEW

**Comment by jacobq** [comment]
Looks like my test isn't passing yet -- marking as WIP for now

**Comment by jacobq** [comment]
It seems that the patch is not being applied at the right time or something because I can see the following in `src/out/Debug/gen/third_party/blink/public/mojom/dom_storage/storage_area.mojom.h`:

```c++
  static constexpr uint32_t kPerStorageAreaQuota = 10485760U;
  
  static constexpr uint32_t kPerStorageAreaOverQuotaAllowance = 102400U;
```

which I suspect is derived from the unpatched `src/content/browser/dom_storage/dom_storage_types.h`

**Comment by jacobq** [comment]
FYI The failing mac test appears to be a fluke (failure was caused by time out and the corresponding code should be completely unrelated to this PR).

**Comment by jacobq** [comment]
Please let me know if there's anything that I can do to help land this fix.

**Comment by zcbenz** [comment]
This PR requires approval from @electron/wg-upgrades.

File: spec/chromium-spec.js [hunk 1]
```diff
@@ -375,16 +375,38 @@ describe('chromium feature', () => {
...
+        it(`throws when attempting to use more than 128MiB in ${storageName}`, () => {
+          expect(() => {
+            const testKeyName = '_electronDOMStorageQuotaStillEnforcedTest'
+            const length = 128 * Math.pow(2, 20) - testKeyName.length
+            try {
+              storage.setItem(testKeyName, 'X'.repeat(length))
+            } finally {
+              storage.removeItem(testKeyName)
+            }
+          })
```
Comment (jacobq) [inline spec/chromium-spec.js]: Did I forget [`.to.throw()`](https://www.chaijs.com/api/bdd/#method_throw) here?

Comment (jacobq) [reply spec/chromium-spec.js]: Looks like I did. I will force push to correct this now.

File: patches/chromium/put_back_deleted_colors_for_autofill.patch [hunk 2]
```diff
@@ -94,7 +94,7 @@ index f1330721a711c61211ea136286ee884bc00e1975..e479fc84f7013984fc53d18cad8812fb
      kColorId_ThrobberSpinningColor,
      kColorId_ThrobberWaitingColor,
 diff --git a/ui/native_theme/native_theme_win.cc b/ui/native_theme/native_theme_win.cc
-index 5b1804538c2e78af23c797004f250369580b4178..97e72bc38a7231ce295ad943e977c69586cc6e07 100644
+index 5b1804538c2e..97e72bc38a72 100644
 --- a/ui/native_theme/native_theme_win.cc
 +++ b/ui/native_theme/native_theme_win.cc
 @@ -673,6 +673,17 @@ SkColor NativeThemeWin::GetSystemColor(ColorId color_id,
```
Comment (nornagon) [inline patches/chromium/put_back_deleted_colors_for_autofill.patch]: Please revert these unrelated changes.

File: patches/chromium/dom_storage_limits.patch [hunk 3]
```diff
@@ -1,80 +1,46 @@
...
-   // shrinking changes to pre-existing files that are over budget.
-   if (check_quota && new_item_size > old_item_size && new_quota_used > quota_)
-     return false;
-+#endif
- 
-   keys_values_.Set(key, value);
-   ResetKeyIterator();
+ // accomodate concurrent writes from different renderers
+diff --git a/third_party/blink/public/mojom/dom_storage/storage_area.mojom b/third_party/blink/public/mojom/dom_storage/storage_area.mojom
+index 1f1b2c6fa109..d8f15eed9be8 100644
```
Comment (nornagon) [inline patches/chromium/dom_storage_limits.patch]: Please use the `git-export-patches` script as described in [the patches documentation](https://github.com/electron/electron/blob/master/docs/development/patches.md#editing-an-existing-patch). It's important that we get the full hashes here, amongst other formatting things that the script does.

File: patches/chromium/put_back_deleted_colors_for_autofill.patch (hunk 2 above)
Comment (jacobq) [reply patches/chromium/put_back_deleted_colors_for_autofill.patch]: Sure, I will drop that commit. (Would you accept a separate PR to fix the "trailing whitespace" warning that gets produced without it?)

File: patches/chromium/dom_storage_limits.patch (hunk 3 above)
Comment (jacobq) [reply patches/chromium/dom_storage_limits.patch]: Ah, I should've read that. (I was just using `git format-patch`) I'll fix that shortly and push again.

File: patches/chromium/put_back_deleted_colors_for_autofill.patch (hunk 2 above)
Comment (nornagon) [reply patches/chromium/put_back_deleted_colors_for_autofill.patch]: sure :)

File: spec/chromium-spec.js [hunk 4]
```diff
@@ -371,16 +371,38 @@ describe('chromium feature', () => {
...
+          // translation between character and byte counts. However,
+          // a string of 40 * 2^20 characters will require at least 40MiB
+          // and presumably no more than 80MiB, a size guaranteed to
+          // to exceed the original 10MiB quota yet stay within the
+          // new 100MiB quota.
+          // Note that both the key name and value affect the total size.
+          const testKeyName = '_electronDOMStorageQuotaIncreasedTest'
+          const length = 40 * Math.pow(2, 20) - testKeyName.length
+          storage.setItem(testKeyName, 'X'.repeat(length))
+          // Wait at least one turn of the event loop to help avoid false positives
```
Comment (nornagon) [inline spec/chromium-spec.js]: What false positives might these be? Ideally we could simplify this test to not require a tick.

Comment (jacobq) [reply spec/chromium-spec.js]: Good question. This might not be applicable any more, but when we were circumventing the restriction prior to this PR, a write followed immediately by a read would "appear to work" (no exceptions throw & reading shows the expected value). Now that we're letting chromium behave more like its usual self (i.e. throwing an exception when a write would exceed the allowed limit) these "false positives" should've be possible any more. Would you like me to remove the `setTimeout`?

Comment (nornagon) [reply spec/chromium-spec.js]: ah, okay. In that case it seems fine to keep the extra tick, it will let us know if things start behaving weirdly again. It'd be great if you could add that context into this comment, though, as it may help the next person to work with this code!

Comment (jacobq) [reply spec/chromium-spec.js]: Updated -- hope that's useful (I'm often wary of comments as they can accidentally be misleading)

**Comment by jacobq** [comment]
Could you give the CI a kick? (lint failed due to "Error: socket hang up")

**Comment by jacobq** [comment]
Looks like win-x64 could use a restart as well... the error looks unrelated:
```
not ok 443 <webview> tag <webview>.capturePage() returns a Promise with a NativeImage
  AssertionError: expected undefined to equal 6
      at Context.<anonymous> (C:\projects\src\electron\spec\webview-spec.js:1035:32)
[4576:1113/203523.801:WARNING:ipc_message_attachment_set.cc(49)] MessageAttachmentSet destroyed with unconsumed attachments: 0/1
```

**Comment by jacobq** [comment]
@deepak1556 I'm sure you're busy, but would you mind taking a quick peek at this and letting me know if you notice anything major that needs to be addressed?

**Comment by jacobq** [comment]
:pray: Thanks, Robo.

**Comment by jacobq** [comment]
> @jacobq has manually backported this PR to "8-x-y", please check out #21380

Hopefully this is the right process.... I just cherry-picked the squashed commit from this onto `8-x-y` since I noticed that this didn't land in the latest beta (v8.0.0-beta.4).
